<img src="hs.png" width="200">

# SCecision trees

## Remark:
In order for this notebook to work propely, you must also download the files to the same folder where you downloaded it (they contain data, you can open them and view them with Notepad or another text editor:
* data.csv
* data_s.csv
* to_classify.csv
* dt.png


## 1. data preparation

Decision trees are on of the simplest and the most intuitive machine learning models. In this notebook, we will introduce you to their functioning using the example of data from a fictitious website for online shopping

We will use the Python 3 language and a copious amount of examples and illustrations. Also we will learn step by step how Python programming code looks like, and we will be glad if this is also an incentive for you to learn Python, which is extremely useful in today's business world.

First we will load the data. To do this, you click on the next cell (where it says "**In[   ]**"), and press *Shift+Enter(Return)*. Wait until it says "**In[\*]**" and until you se "**In[1]**", then the calculation is complete. As a result you should get the first five rows of the table.

In [ ]:
import pandas as pd
data = pd.read_csv("data.csv")
podatci.head()

**Task 1:** We see that this short script (or program, or applcation) lists the first five lines from the table  "algebra.csv". In the cell below, try modifying the cript to list the firts ten lines:

In [2]:
data.head(5)#modify this line to get 10 lines

customer_id  age  prev_purchases_in_usd gender  uses_whishlist  item_id  \
0         1000   34                1234.55      f               1      234   
1         1001   23                   0.00      m               1      257   
2         1002   44                   0.00      f               0      222   
3         1922   35                   0.00      f               0      258   
4         1004   34                 900.99      m               1      334   

   price category end_result  
0  89.99     tech          3  
1  13.45     tech          2  
2  54.99     book          3  
3  11.45     tech          2  
4  99.99     tech          2

## 2. The idea behind the classifier

A decision tree is conceptually the simplest machine learning algorithm, so we will show how machine learning works on it. 

Machine learning algorithms can be roughly divided into  *classifocation* algorithms and *clustering* algorithms . A decision tree is a classification algorithm.

Classification generally works like this. We declare one column from the data table as  the **targeted** variable. In our case is the  *end_result* column, but it can be any other column that we are interested in  (that is, for which we want to find out how the values in the other columns  *predicts* it)

Let's raise the first three columns of our table( press *Shift+Enter* on the next cell):

In [ ]:
data.head(3)

The meaning of the *end_result* is next:
* Value 0 means "abandoned cart"
* Value 1 means "unfinished payment"
* Value 2 means "card declined during payment"
* Value 3 means "Successful payment"

Let's imagine for now these three lines. If our target variable is *end_result*, then it means that we want to predict the value of  *end_result* based on the results in other columns. For example, if we want to predict who abandoned the cart, we see that it is true on these three lines "If **gender = m**, then **end_result = 2**".

**gender = m** is called a *predictor* for target variable **end_result** (that is,to recognize when  **end_result = 2**). Notice that **gender = m** is a predictor, while  **gender** itself is called *property* odnosno *feature* 

There can be more predictors, and in this example we see that there are more:

* **gender = m**
* **age < 28** (also it could have been 24, 25, 26,...,32, but we chose 28 because it is in the middle)
* **item_id > 250** (again , it can be any value that separates them)
* Can you find another? (hint: price)

We also note that predictors can be complex. For example, the predictor **prev_purchases_in_usd = 0 AND uses_whishlist = 1** also works well. This predictor also points to  **end_result = 2**.

As a small digression, how would we formulate a predictor using the feature **prev_purchases_in_usd = 0** i **uses_whishlist = 1** for **end_result = 2**?

The answer is **prev_purchases_in_usd > 0 OR uses_whishlist = 0** (notice the changes, AND in OR, and values).

But let's go back to decision trees. Let's load the following table row:

In [ ]:
data.head(4)

Now **gender = m**, **age < 28**, **item_id > 0** are no longer good predictors,nor is our complex predictor. The price based predictor still works. Let's load another table row:


In [ ]:
data.head(5)

In theory, we could manually construct the predictors to catch the correct cases. For example  **price < 20 AND price > 95**, or even better we would take **customer_id = 1001 OR customer_id = 1922 OR customer_id = 1004**. This would lead us in the totally wrong direction. 

We wany yo create a tool that will predict for future customers whether they will abandon or not, and thw method of counting by **customer_id** (means focusing only on this column without thinking about other features) who has abandoned so far will certainly nt help us to win a client **customer_id = 5555**, neither to keep the customer **customer_id = 0003**.


## 3. Classification and marketing

Here it is appropriate to give one important maxim:

_**Acquisition of new clients and retention of existing clients are the two main tasks of marketing**._ 

The emphasis may be a little more on one or the other, but this is the *essence* of marketing. 

We want to "learn" from the data the cobination of properties shared by customers who have given up, and then, based on that combinatiom, try to see what our problem is, but also predict whether customer 5555 will have the same problem, and thus be able to act preventively on his retention. In practice, retention is the most common way to eliminate difficulties for clients, beacuse the main reason for leaving (which we can act on) is some problem with our system. Of course, there are other reasons, but they are (a) significantly more difficult to find, and (b) significantly more difficult to act on.

Another aspect of this analysis is to see where there is room for acqusition. The asquisition is simpler on the one hand, and more complex on the other.

It is easier to find the best predictor for  **end_result = 3**, and analyze the result. With a little luck, among the best results we will have valuable insights, eg. **age > 25 AND age < 40 AND gender = f**, with which we can then run very good targeted campaigns.

It is more complex in that the acquisition can also be viewed in the way of  *cross-selling*, that is, if we have learned from our data that clients who buy any book for over 70 USD most often also buy item 258, then anyone who puts in basket a book over 70 USD  we want to offer and an item 258 (maybe with a discount) 


## 4. Decision trees

Decision trees select features and combine them so that a specific oreder of their application is obtained. The main trick is to measure which features give the best disicirmination, and take them first. A good example is predicting who did (and who would) survive on the  _Titanic_ (borrowing an image from Wikipedia)

<img src="dt.png">

The numbers below the elements of the image indicate the probability of survival (decimal numbers of the form 0.34), and the percentage of the tatal caeses that are in that category (Which, for the sake of rounding, add up to 101%, but this is not important noe). A probability of survival above 0.5 is considered the "survived" category, and below 0.5 is considered the "died" category (this technique where the numerical scale is divided into two categories before and after a ceryain limit is called "thresholding").

So, if a person is female, her probability of survival is 0.73 (73%). If he is a male and older than 9.5 years, then he probably died (survival probability  0.17, i.e. 17%). Male persons younger than 9.5 years who have less than 2.5 brothers and sisters had the best chance of survival (0.89, or 89% chance, although  there were very few of them, or 2% of the total population). This 2% means that there were quite a few samples with this predictor (**sex = m AND age < 9.5 AND sibsp < 2.5**), which means that perhaps this classification is not the most reliable, that is, it may be too dependent on the chance.

Classification based on gender gives a 0.73 chance of survival, but this has the ability to easily (with only one feature **sex**) capture 36% of the overall data. In a certain sense, it is the strongest feature and the most stable (which also follows common sense intuition). 

This is the basis of the functioning of decision trees: the decision tree tries to find an order by delimiting first by the strongest feature, and then by the weaker feature. 

This does **not** mean that the strongest feature is the one that gives the greatest chance of survival (that is, the biggest difference regarding *thresholdinga*), because that is no longer importnant, what matters is whether it is above 0.5, so it gets the lable "SURVIVABLE" or below, so it gets the lable "DYING". 

The strongest feature is the one that best divides the samples, so in our case 36% vs 64%. The details of how the tree calculates exactly which oreder it will go in are complex ( and that's exactly why we use an algorithm and not do this manually), but this description is the intuition behind it.

Now we are ready to build a decision tree over our data. The next few cells shoul be executed with Shift+Enter, and we give explanations below them.

In [3]:
import pandas as pd
from sklearn import tree

A quick comment:

* The first line serves to load the additional module "Pandas" which serves as a table (similar to Excel),
* The second line loads an additional Python module (sklearn) zfor its submodule (tree) for decision trees,

The next step is to load the data(Shift+Enter to the bottom cell):

In [5]:
data_2 = pd.read_csv("data_s.csv")

**NOTE:** this is not the same table as the one above, but the number of columns has been reduced (the rows are the same). Execute the cell below to see the first few rows:

In [ ]:
data_2.head()

Column **customer_id** is not needed anyomre,so we are going to remove it:

In [6]:
data_3 = data_2.drop('customer_id',axis=1)
data_3.head()#this line is overlooking if did everything fine

age  price  end_result
0   34  89.99           3
1   23  13.45           2
2   44  54.99           3
3   35  11.45           2
4   34  99.99           2

The next step is to split the data into two tables, one that has a feature and the other that has  **end_result** in it (we do this because that's how SKlearn wants to recive data):

In [7]:
feature = data_3[["age","price"]]
feature.head()

age  price
0   34  89.99
1   23  13.45
2   44  54.99
3   35  11.45
4   34  99.99

In [8]:
results = data_3[["end_result"]]
results.head()

end_result
0           3
1           2
2           3
3           2
4           2

Let's revise that "head" means to print the first few results (default=5, if you want some other value, you should wirte it in brackets), and if you want to see the whole table, just write "results" instead of "results.head()".

In [9]:
tree = tree.DecisionTreeClassifier()
Breza = tree.fit(features, results)

The first line in the cell above creates a new empty tree from the SKlearn module, and the second line trains it over the data from the CSV table we explored. Now the tree is trained and lives in the system under the name "Breza" and is ready to classify new data.

Let's imagine that we now get a table of customers who are currently on the page and buying, and that is the "to_classify.csv". Let's lead that table  (Shift+Enter to the bottom cell)

In [10]:
cleints_on_the_site = pd.read_csv("to_classify.csv") #this prints the table
clients_on_the_site

age   price
0   23   45.67
1   56   23.45
2   12   99.99
3   34   23.45
4   56   89.67
5   35   56.78
6   28  123.45
7   70   14.67
8   67   99.67
9   50   50.00

Now let's run Breza over this data:

In [11]:
pred = Breza.predict(clients_on_the_site)
pred

array([2, 1, 0, 2, 3, 3, 3, 1, 0, 3], dtype=int64)

In this way, Breza wrote down what each of the clients would most likely do.
Let's remember:

* Value 0 means "abandoned cart"
* Value 1 means "unfinished payment"
* Value 2 means "card desclined during payment"
* Value 3 means "Successful payment"

We can add this data to the table in the new  **expected_action**:

In [ ]:
clients_on_the_site['expected_action'] = pred
clients_on_the_site